File: schema_salad/java/main_utils/UriLoader.java
package ${package}.utils;

import java.util.ArrayList;
import java.util.List;

public class UriLoader<T> implements Loader<T> {
  private final Loader<T> innerLoader;
  private final boolean scopedId;
  private final boolean vocabTerm;
  private final Integer scopedRef;

  public UriLoader(
      final Loader<T> innerLoader,
      final boolean scopedId,
      final boolean vocabTerm,
      final Integer scopedRef) {
    this.innerLoader = innerLoader;
    this.scopedId = scopedId;
    this.vocabTerm = vocabTerm;
    this.scopedRef = scopedRef;
  }

  private Object expandUrl(
      final Object object, final String baseUri, final LoadingOptions loadingOptions) {
    if (object instanceof String) {
      return loadingOptions.expandUrl(
          (String) object, baseUri, this.scopedId, this.vocabTerm, this.scopedRef);
    } else {
      return object;
    }
  }

  public T load(
      final Object doc_,
      final String baseUri,
      final LoadingOptions loadingOptions,
      final String docRoot) {
    Object doc = doc_;
    if (doc instanceof List) {
      List<Object> docList = (List<Object>) doc;
      List<Object> docWithExpansion = new ArrayList<Object>();
      for (final Object el : docList) {
        docWithExpansion.add(this.expandUrl(el, baseUri, loadingOptions));
      }
      doc = docWithExpansion;
    }
    if (doc instanceof String) {
      doc = this.expandUrl(doc, baseUri, loadingOptions);
    }
    return this.innerLoader.load(doc, baseUri, loadingOptions);
  }
}

File: schema_salad/dotnet/util/Loaders/UriLoader.cs
using System.Collections;

namespace ${project_name};

internal class UriLoader : ILoader<object>
{
    readonly ILoader inner;
    readonly bool scopedID;
    readonly bool vocabTerm;
    readonly int? scopedRef;

    public UriLoader(in ILoader inner, in bool scopedID, in bool vocabTerm, in int? scopedRef)
    {
        this.inner = inner;
        this.scopedID = scopedID;
        this.vocabTerm = vocabTerm;
        this.scopedRef = scopedRef;
    }

    public object Load(in object doc_, in string baseuri, in LoadingOptions loadingOptions, in string? docRoot = null)
    {
        object doc = doc_;
        if (doc is IList)
        {
            List<object> docList = (List<object>)doc_;
            List<object> docWithExpansion = new();
            foreach (object val in docList)
            {
                if (val is string valString)
                {
                    docWithExpansion.Add(loadingOptions.ExpandUrl(valString, baseuri, scopedID, vocabTerm, scopedRef));
                }
                else
                {
                    docWithExpansion.Add(val);
                }
            }

            doc = docWithExpansion;
        }
        else if (doc is string docString)
        {
            doc = loadingOptions.ExpandUrl(docString, baseuri, scopedID, vocabTerm, scopedRef);
        }

        return (object)inner.Load(doc, baseuri, loadingOptions);
    }

    object ILoader.Load(in object doc, in string baseuri, in LoadingOptions loadingOptions, in string? docRoot)
    {
        return Load(doc,
                    baseuri,
                    loadingOptions,
                    docRoot)!;
    }
}
